pkg/controllers/management/auth/globalroles/fleetworkspace_role_handler.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/rancher/rancher/pkg/types/config"
+	rbacv1 "github.com/rancher/wrangler/v2/pkg/generated/controllers/rbac/v1"
+	v1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	localFleetWorkspace            = "fleet-local"
+	fleetWorkspaceClusterRulesName = "fwcr"
+	fleetWorkspaceVerbsName        = "fwv"
+)
+
+// fleetWorkspaceRoleHandler manages ClusterRoles created for the InheritedFleetWorkspacePermissions field. It manages 2 roles:
+// - 1) CR for ResourceRules
+// - 2) CR for WorkspaceVerbs
+type fleetWorkspaceRoleHandler struct {
+	crClient rbacv1.ClusterRoleController
+	crCache  rbacv1.ClusterRoleCache
+	fwCache  mgmtcontroller.FleetWorkspaceCache
+}
+
+func newFleetWorkspaceRoleHandler(management *config.ManagementContext) *fleetWorkspaceRoleHandler {
+	return &fleetWorkspaceRoleHandler{
+		crClient: management.Wrangler.RBAC.ClusterRole(),
+		crCache:  management.Wrangler.RBAC.ClusterRole().Cache(),
+		fwCache:  management.Wrangler.Mgmt.FleetWorkspace().Cache(),
+	}
+}
+
+// ReconcileFleetWorkspacePermissions reconciles backing ClusterRoles created for granting permission to fleet workspaces.
+func (h *fleetWorkspaceRoleHandler) reconcileFleetWorkspacePermissions(gr *v3.GlobalRole) error {
+	if err := h.reconcileResourceRules(gr); err != nil {
+		return fmt.Errorf("error reconciling fleet permissions cluster role: %w", err)
+	}
+	if err := h.reconcileWorkspaceVerbs(gr); err != nil {
+		return fmt.Errorf("error reconciling fleet workspace verbs cluster role: %w", err)
+	}
+
+	return nil
+}
+
+func (h *fleetWorkspaceRoleHandler) reconcileResourceRules(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceClusterRulesName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
+
+	if apierrors.IsNotFound(err) {
+		_, err := h.crClient.Create(backingResourceRulesClusterRole(gr, crName))
+
+		return err
+	}
+
+	if !equality.Semantic.DeepEqual(cr.Rules, gr.InheritedFleetWorkspacePermissions.ResourceRules) {
+		// undo modifications if cr has changed
+		cr.Rules = gr.InheritedFleetWorkspacePermissions.ResourceRules
+		_, err := h.crClient.Update(cr)
+
+		return err
+	}
+
+	return err
+}
+
+func (h *fleetWorkspaceRoleHandler) reconcileWorkspaceVerbs(gr *v3.GlobalRole) error {
+	crName := wrangler.SafeConcatName(gr.Name, fleetWorkspaceVerbsName)
+	cr, err := h.crCache.Get(crName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	crMissing := apierrors.IsNotFound(err)
+
+	if gr.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil {
+		return h.deleteClusterRoleIfNeeded(!apierrors.IsNotFound(err), crName)
+	}
```

Comment:
Similar to the above, we check if the error is not found 3 times here. I would recommend re-writing this to minimize the amount of checking we do for this value.

Reply:
I refactored a bit this part too. However, in this case we have to check more times the `isNotFound` error. That's why I added the `crMissing` var. The reason for that is that I want to prevent listing all fleetworkspaces if `InheritedFleetWorkspacePermissions.WorkspaceVerbs` is nil, which will happens for all existing `GlobalRoles` that doesn't have `InheritedFleetWorkspacePermissions`